src/categorical_algebra/Slices.jl

```diff
@@ -0,0 +1,133 @@
+module Slices
+export Slice, SliceMorphism,SliceLimit,SliceColimit, pushout_complement
+
+using AutoHashEquals
+using ...GAT
+using ..FreeDiagrams, ..Limits, ..CSets
+using ...Theories: Category
+import ...Theories: dom, codom, compose, id
+import ..Limits: limit, colimit, universal
+import ..FinSets: force, pushout_complement
+
+@auto_hash_equals struct Slice{Ob,Hom}
+  slice::Hom
+end
+
+@auto_hash_equals struct SliceMorphism{Ob, Hom}
+  dom::Slice{Ob, Hom}
+  codom::Slice{Ob, Hom}
+  f::Hom
+  function SliceMorphism{Ob, Hom}(dom, codom, f) where {Ob, Hom}
+    p1, p2 = force.([dom.slice, compose(f,codom.slice)])
+    p1 == p2 ||
+       error("Slice condition failed: \ndom.slice $p1\ncomposite $p2\nf " *
+             "$(force(f))\ncodom $(force(codom.slice))")
+    return new{Ob,Hom}(dom, codom, f)
+  end
+end
+
+SliceMorphism(a,b,f) = SliceMorphism{typeof(a).parameters...}(a,b,f)
+
+
+dom(s::Slice)::StructACSet = dom(s.slice)
+codom(s::Slice)::StructACSet = codom(s.slice)
+force(s::Slice{Ob,Hom}) where {Ob,Hom} = Slice{Ob,Hom}(force(s.slice))
```

Comment:
Since calling `force` can change the concrete type of its argument (e.g., from `SetFunctionCallable` to `FinDomFunctionVector`), this should be simply:

```julia
force(s::Slice) = Slice(force(s.slice))
```